src/zjs_promise.c

```diff
@@ -5,155 +5,596 @@
+    // 7.
+    return trigger_promise_reactions(reactions, value);
+}
+
+// 25.4.1.3.1
+static jerry_value_t reject_function(const jerry_value_t function_obj,
+                                     const jerry_value_t this,
+                                     const jerry_value_t argv[],
+                                     const jerry_length_t argc)
+{
+    ZVAL promise = zjs_get_property(function_obj, "Promise");
+    bool already_resolved;
+    zjs_obj_get_boolean(function_obj, "AlreadyResolved", &already_resolved);
+    if (already_resolved == true) {
+        return ZJS_UNDEFINED;
     }
+    zjs_obj_add_boolean(function_obj, true, "AlreadyResolved");
 
-    return jerry_acquire_value(this);
+    return do_reject_promise(promise, argv[0]);
 }
 
-static jerry_value_t promise_then(const jerry_value_t function_obj,
-                                  const jerry_value_t this,
-                                  const jerry_value_t argv[],
-                                  const jerry_length_t argc)
+// 25.4.1.3.2
+static jerry_value_t resolve_function(const jerry_value_t function_obj,
+                                      const jerry_value_t this,
+                                      const jerry_value_t argv[],
+                                      const jerry_length_t argc)
 {
-    return promise_resolve(function_obj, this, argv, argc, 1);
+    // 2.
+    ZVAL promise = zjs_get_property(function_obj, "Promise");
+    if (!jerry_value_is_object(promise)) {
+        ERR_PRINT("promise is not object\n");
+    }
+
+    // 3.
+    bool already_resolved;
+    zjs_obj_get_boolean(function_obj, "AlreadyResolved", &already_resolved);
+    // 4.
+    if (already_resolved == true) {
+        return ZJS_UNDEFINED;
+    }
+    // 5.
+    zjs_obj_add_boolean(function_obj, true, "AlreadyResolved");
+    // 6.
+    if (promise == argv[0]) {
+        return do_reject_promise(promise, TYPE_ERROR("resolution == promise"));
+    }
+    // 7.
+    if (argc < 1 ) {
+        return do_fulfill_promise(promise, ZJS_UNDEFINED);
+    } else if (!jerry_value_is_object(argv[0])) {
+        return do_fulfill_promise(promise, argv[0]);
+    }
+    // 8.
+    ZVAL then = zjs_get_property(argv[0], "then");
+    // 9. // abrupt completion???
+    // 10. thenAction = then
+    // 11.
+    if (!jerry_value_is_function(then)) {
+        return do_fulfill_promise(promise, argv[0]);
+    }
+    // 12. // EnqueueJob
+    promise_thenable_job(promise, argv[0], then);
+    // 13.
+    return ZJS_UNDEFINED;
 }
 
-static jerry_value_t promise_catch(const jerry_value_t function_obj,
-                                   const jerry_value_t this,
-                                   const jerry_value_t argv[],
-                                   const jerry_length_t argc)
+// 25.4.1.3
+static jerry_value_t create_resolving_functions(jerry_value_t promise)
 {
-    return promise_resolve(function_obj, this, argv, argc, 0);
+    ZVAL already_resolved = jerry_create_boolean(false);
```

Comment:
i think the already_resolved should be an boolean object instead of a boolean value.
both resolve.[[AlreadyResolved]] and reject.[[AlreadyResolved]] will ref to same object